Bump to OSSM 3.0 for Gateway API support

Miciah opened the pull request:
OSSM 3.0 replaces the OpenShift Service Mesh Operator [with Sail Operator](https://developers.redhat.com/articles/2023/08/30/istio-sail-operator-openshift), and the ServiceMeshControlPlane CRD is replaced by the Istio CRD.  Vendor the sail-operator API:

    go mod edit -replace github.com/imdario/mergo=github.com/imdario/mergo@v0.3.5
    go get github.com/istio-ecosystem/sail-operator/api/v1alpha1
    go mod tidy
    go mod vendor

Note that vendoring Sail Operator requires the mergo override.

Update the gatwayclass controller to create a subscription for Sail Operator instead of the Service Mesh Operator, and to create an Istio CR instead of a ServiceMeshControlPlane CR.

Currently, [OSSM 3.0 is Tech Preview](https://www.redhat.com/en/blog/red-hat-openshift-service-mesh-3-now-technology-preview), so we must configure the subscription accordingly.

Miciah commented:
/test e2e-aws-gatewayapi

Miciah commented:
/test e2e-aws-gatewayapi

candita commented:
@Miciah Just fyi - this PR doesn't update the CRDs so this is still running v1beta1 Gateway API CRDs.

candita commented:
/assign @candita, @alebedev87

candita commented:
/assign @candita @alebedev87

Miciah commented:
/test e2e-aws-gatewayapi
/test e2e-aws-operator

Miciah commented on pkg/operator/controller/gatewayclass/subscription.go:
```diff
@@ -55,11 +55,12 @@ func desiredSubscription(name types.NamespacedName) (*operatorsv1alpha1.Subscrip
-			Channel:                "stable",
+			Channel:                "1.1-nightly", //"stable",
 			InstallPlanApproval:    operatorsv1alpha1.ApprovalAutomatic,
-			Package:                "servicemeshoperator",
-			CatalogSource:          "redhat-operators",
+			Package:                "sailoperator",        //"servicemeshoperator3",
+			CatalogSource:          "community-operators", //"redhat-operators",
 			CatalogSourceNamespace: "openshift-marketplace",
+			StartingCSV:            "sailoperator.v1.1.0-nightly-2025-03-05",
```
OSSM 3.0.0TP2 still has the Istio v1alpha1 CRD, so I am using a Sail Operator nightly to get the Istio v1 CRD.  I will update this subscription once OSSM 3.0 is GA.

Miciah commented:
I forgot to `git add` a couple of new files.  https://github.com/openshift/cluster-ingress-operator/compare/fbba0d2acaa26ba4fea21233ba472426885945e7..3e738c1b36c922bcdac731effce285edbe8050d0 adds the missing files.

/test e2e-aws-gatewayapi
/test e2e-aws-operator

Miciah commented:
/test e2e-aws-gatewayapi

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/compare/3e738c1b36c922bcdac731effce285edbe8050d0..4e54863f46ae056047b6b18730caf0502c139b90 has some E2E test fixes.

Miciah commented:
I believe that the tests are failing because automated deployment/sidecar injection is not working with Sail Operator.

Miciah commented:
/test e2e-aws-gatewayapi

Miciah commented:
[e2e-aws-gatewayapi failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1152/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-gatewayapi/1898094066692788224) because `Step e2e-aws-gatewayapi-aws-deprovision-users-and-policies failed`.

/test e2e-aws-gatewayapi

guicassolato commented:
You may be able to ditch that `replace` directive on `github.com/imdario/mergo` (as well as the dep altogether) if you can bump your `k8s.io/*` deps to v0.32.2+, `k8s.io/client-go` in particular.

Ref.: https://github.com/kubernetes/kubernetes/pull/126764

Miciah commented:
> You may be able to ditch that `replace` directive on `github.com/imdario/mergo` (as well as the dep altogether) if you can bump your `k8s.io/*` deps to v0.32.2+, `k8s.io/client-go` in particular.
> 
> Ref.: [kubernetes/kubernetes#126764](https://github.com/kubernetes/kubernetes/pull/126764)

Alas, I get the same issue even with k8s.io/* v0.32.2.

I used the following commands:

    go get k8s.io/api@v0.32.2 k8s.io/apiextensions-apiserver@v0.32.2 k8s.io/apimachinery@v0.32.2 k8s.io/apiserver@v0.32.2
    go mod edit -replace k8s.io/client-go=k8s.io/client-go@v0.32.2
    go mod tidy
    go mod vendor
    go get github.com/istio-ecosystem/sail-operator/api/v1

The last command still complains about the mergo package:

    go: downloading github.com/istio-ecosystem/sail-operator v0.0.0-20250311054604-76c2b6706812
    go: github.com/istio-ecosystem/sail-operator@v0.0.0-20250311054604-76c2b6706812 (matching github.com/istio-ecosystem/sail-operator/api/v1@upgrade) requires github.com/imdario/mergo@v1.0.0: parsing go.mod:
    	module declares its path as: dario.cat/mergo
    	        but was required as: github.com/imdario/mergo

I guess we need sail-operator to likewise drop its dependency on mergo before I can do away with the replace rule.

Miciah commented:
[e2e-aws-gatewayapi failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1152/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-gatewayapi/1899505646944915456) because of an ongoing CI outage:
```text
   * failed to generate integrated streams: failed to get integrated stream ocp/4.19: failed to make request to configresolver: Get "https://config.ci.openshift.org/integratedStream?name=4.19&namespace=ocp": GET https://config.ci.openshift.org/integratedStream?name=4.19&namespace=ocp giving up after 6 attempt(s) 
```

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/compare/40866c7e8e4fdaf156d9f4456c002a7ba84161c0..ce07ee6f3705d67ba11bdd953f9f0be05da03311 does the following:

- Bump from Istio v1.24.2 to v1.24.3.  
- Set `PILOT_MULTI_NETWORK_DISCOVER_GATEWAY_API=false`.  
- Add logging in `assertGatewaySuccessful` to aid in diagnosing test failures.

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/compare/ce07ee6f3705d67ba11bdd953f9f0be05da03311..54318e9606a1d5a9f42c5c73197c352a343ea68a rebases for #1192.

candita commented:
No IP Pool leases left for last test:

> [36mINFO[0m[2025-03-12T20:35:55Z] Acquiring leases for test e2e-aws-gatewayapi: [aws-quota-slice] 
> [36mINFO[0m[2025-03-12T20:35:55Z] Acquired 1 lease(s) for aws-quota-slice: [us-west-2--aws-quota-slice-29] 
> [36mINFO[0m[2025-03-12T20:35:55Z] Acquiring IP Pool leases for test e2e-aws-gatewayapi: aws-ip-pools-us-west-2 
> [36mINFO[0m[2025-03-12T20:35:55Z] no leases of type: aws-ip-pools-us-west-2 available 
>

/test e2e-aws-gatewayapi

candita commented:
#1198 makes it so the e2e-aws-gatewayapi test can't fully run anymore.

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/compare/5b23d1061d3333e94800da928a6ffd5e5abe37c8..30b576c4293b2962b137bcb93ff77f278d7bf681 makes the following changes:

- Evidently the label key `istio.io/gateway-name` was changed to `gateway.networking.k8s.io/gateway-name`; update the gateway-service-dns controller accordingly.  
- Added the `istio.io/rev` label to the openshift-ingress namespace.  This is a temporary solution; we need to figure out how to make this label not required.  
- In the `ensureCRDs` test helper, report all served versions, not just the last listed one.
- In the `createGatewayClass` test helper, correctly handle errors getting an existing gatewayclass.

Miciah commented:
/test e2e-aws-operator-techpreview

lihongan commented on hack/gatewayapi-conformance.sh:
```diff
@@ -47,7 +47,7 @@ go mod vendor
 # because the AWS ELB needs extra ~60s for DNS propagation
 sed -i "s/MaxTimeToConsistency:              30/MaxTimeToConsistency:              90/g" conformance/utils/config/timeout.go
```
I noticed that one more space is inserted in release-1.1 and later, the workaround could be updated
```suggestion
sed -i "s/MaxTimeToConsistency:.*30/MaxTimeToConsistency: 90/g" conformance/utils/config/timeout.go
gofmt -s -w conformance/utils/config/timeout.go
```

Miciah replied:
Thanks!  Fixed in https://github.com/openshift/cluster-ingress-operator/pull/1152/commits/26a3918c7374cc16eac04b87c9e4267591c304a5.

candita commented on hack/gatewayapi-conformance.sh:
```diff
@@ -47,7 +47,7 @@ go mod vendor
 sed -i "s/MaxTimeToConsistency:              30/MaxTimeToConsistency:              90/g" conformance/utils/config/timeout.go
 
-SUPPORTED_FEATURES="Gateway,HTTPRoute,ReferenceGrant,GatewayPort8080,HTTPRouteQueryParamMatching,HTTPRouteMethodMatching,HTTPRouteResponseHeaderModification,HTTPRoutePortRedirect,HTTPRouteSchemeRedirect,HTTPRoutePathRedirect,HTTPRouteHostRewrite,HTTPRoutePathRewrite,HTTPRouteRequestMirror,HTTPRouteRequestMultipleMirrors,HTTPRouteBackendProtocolH2C,HTTPRouteBackendProtocolWebSocket"
+SUPPORTED_FEATURES='Gateway,HTTPRoute,ReferenceGrant,GatewayHTTPListenerIsolation,GatewayInfrastructurePropagation,GatewayPort8080,GatewayStaticAddresses,HTTPRouteBackendProtocolH2C,HTTPRouteBackendProtocolWebSocket,HTTPRouteBackendRequestHeaderModification,HTTPRouteBackendTimeout,HTTPRouteDestinationPortMatching,HTTPRouteHostRewrite,HTTPRouteMethodMatching,HTTPRouteParentRefPort,HTTPRoutePathRed ...
```
Some of these features look like extended features, which we don't need to test. It would save some time not to test them.  We should only need to test the standard features, excluding mesh.  That would be any feature at the code links below that is listed in the `FeatureChannelStandard`. 

Gateway: https://github.com/kubernetes-sigs/gateway-api/blob/main/pkg/features/gateway.go#L65-L86

HTTPRoute: https://github.com/kubernetes-sigs/gateway-api/blob/main/pkg/features/httproute.go#L103-L193 

GRPCRoute: https://github.com/kubernetes-sigs/gateway-api/blob/main/pkg/features/grpcroute.go#L31-L34

ReferenceGrant: https://github.com/kubernetes-sigs/gateway-api/blob/main/pkg/features/referencegrant.go#L31-L34

Miciah replied:
Thanks!  Fixed in https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5.  I still get some failures, though, and the test takes over an hour to run (it didn't finish before my cluster expired).

candita commented on manifests/00-cluster-role.yaml:
```diff
@@ -182,9 +182,9 @@ rules:
+  - sailoperator.io
   resources:
-  - servicemeshcontrolplanes
+  - istios
```
And istioRevisions ?

Miciah replied:
Hm, I hadn't thought of that.  However, cluster-ingress-operator doesn't use istiorevisions itself, so it shouldn't need the permission.

candita commented on manifests/00-cluster-role.yaml:
```diff
@@ -182,9 +182,9 @@ rules:
 
 - apiGroups:
-  - maistra.io
+  - sailoperator.io
```
Does this need to be updated, or are the servicemesh3 groups in sailoperator.io?

Miciah replied:
OSSM 3 uses the sailoperator.io API group.

candita commented on pkg/operator/client/client.go:
```diff
@@ -10,8 +10,7 @@ import (
 
-	maistrav1 "github.com/maistra/istio-operator/pkg/apis/maistra/v1"
-	maistrav2 "github.com/maistra/istio-operator/pkg/apis/maistra/v2"
+	sailv1 "github.com/istio-ecosystem/sail-operator/api/v1"
```
Does this need update to match up with the servicemesh3 operator?

Miciah replied:
I don't think so.  Since OSSM and the upstream Sail Operator use the same GVK, there shouldn't be any breaking changes.  I see some minor differences between them, mostly around kubebuilder validation (which doesn't matter when we're just importing the Go definitions) and some API fields that we aren't using anyway.

candita commented on pkg/operator/controller/gateway-service-dns/controller.go:
```diff
@@ -42,7 +42,7 @@ const (
 	// this label in the selector of any service that it creates for a
 	// gateway.
-	gatewayNameLabelKey = "istio.io/gateway-name"
+	gatewayNameLabelKey = "gateway.networking.k8s.io/gateway-name"
```
Is this an OSSM3 change?

Miciah replied:
Yes.  I don't know whether it's OSSM itself or Istio, but the change came with the bump from OSSM 2 to OSSM 3.

candita commented on pkg/operator/controller/gatewayclass/controller.go:
```diff
@@ -131,17 +131,19 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
-		r.startSMCPWatch.Do(func() {
-			isOurSMCP := predicate.NewPredicateFuncs(func(o client.Object) bool {
-				return o.GetName() == operatorcontroller.ServiceMeshControlPlaneName(r.config.OperandNamespace).Name
+		// The Sail operator installs the istios.sailoperator.io CRD.
```
Change to reference OSSM3 instead of Sail?

Miciah replied:
I don't think it matters for code comments.  As developers, we know that the OSSM operator is the Sail Operator.  I changed it for consistency, though: https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5

candita commented on pkg/operator/controller/gatewayclass/istio.go:
```diff
@@ -0,0 +1,209 @@
+	"github.com/google/go-cmp/cmp/cmpopts"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+	sailv1 "github.com/istio-ecosystem/sail-operator/api/v1"
```
Sailv1?

Miciah replied:
Yeah, it's the v1 version of the Istio API from the sailoperator.io API group, so it makes sense to import it as sailv1.  I think it would be confusing to import it as something else.  The downstream version of the operator modifies the controller logic, but it's the same API (by and large).

candita replied:
My question here and elsewhere is why are we using github.com/istio-ecosystem/sail-operator/api/v1 (upstream repo) rather than github.com/openshift-service-mesh/sail-operator/api/v1 (downstream).

candita commented on pkg/operator/controller/gatewayclass/istio.go:
```diff
@@ -0,0 +1,209 @@
+					IngressControllerMode: sailv1.MeshConfigIngressControllerModeOff,
+				},
+			},
+			Version: "v1.24.3",
```
Would it be worthwhile for us to conduct a brief review of the choice of this particular version?  Do we have the option of using a different version with servicemeshoperator3?

Miciah replied:
I just picked the version that I knew OSSM 3 supported.  OSSM 3.0.0 [supports Istio 1.24.3](https://github.com/openshift-service-mesh/sail-operator/blob/3.0.0/api/v1/istio_types.go#L39-L40), which [has Gateway API v1.2.1](https://github.com/istio/istio/blob/1.24.3/go.mod#L108).  (Istio 1.24.2 [has some prerelease version](https://github.com/istio/istio/blob/1.24.2/go.mod#L108).)

candita commented on pkg/operator/controller/gatewayclass/istio.go:
```diff
@@ -0,0 +1,209 @@
+			Version: "v1.24.3",
+		},
+	}
+	return &istio, nil
```
If there's no possibility of an error can the return value be removed?

candita commented on pkg/operator/controller/gatewayclass/istio.go:
```diff
@@ -0,0 +1,209 @@
+
+	// Diff before updating because the client may mutate the object.
+	// Only diff spec because status may include unexported fields that
+	// cause go-cmp to panic.
```
I wonder if there are labels/annotations that don't want to compare as well, e.g. if the istiod/proxy reconciles add labels/annotations that we are supposed to ignore?

candita commented on pkg/operator/controller/names.go:
```diff
@@ -22,6 +22,10 @@ const (
 
+	// IngressOperatorOwnedAnnotation is the key for the annotation that
+	// indicates that the ingress operator owns the annotated resource.
+	IngressOperatorOwnedAnnotation = "ingress.operator.openshift.io/owned"
```
Why is this needed?

Miciah replied:
There was a suggestion on a recent working-group call that we do something to help us distinguish in the future whether the subscription was created by the operator or by the cluster-admin (or some other entity).  This might turn out to be critical to the OSSM upgrade logic for OCP 4.19→4.20.  Anyway, it doesn't hurt to have this annotation.

candita replied:
nit: It would be nice to ensure that future developers don't use this annotation on a wider basis because they're unsure why it's here.  Could you add a comment that it is really only supposed to be used by subscriptions?

Miciah replied:
Sorry, I missed this comment before.  https://github.com/openshift/cluster-ingress-operator/pull/1152/commits/5c76a8511cee19aeec606b2dffd70d97cf595828 expands the comment to say that the annotation is currently only intended to be applied to newly created subscriptions.

candita commented on pkg/operator/controller/gatewayclass/subscription.go:
```diff
@@ -53,11 +53,14 @@ func desiredSubscription(name types.NamespacedName) (*operatorsv1alpha1.Subscrip
 			Name:      name.Name,
+			Annotations: map[string]string{
+				operatorcontroller.IngressOperatorOwnedAnnotation: "",
+			},
```
Do we have to add this new annotation to function `routerNamespaceChanged` in pkg/operator/controller/ingress/namespace.go?

candita commented on pkg/operator/controller/names.go:
```diff
@@ -269,23 +273,21 @@ func IngressClassName(ingressControllerName string) types.NamespacedName {
 	return types.NamespacedName{
 		Namespace: "openshift-operators",
-		Name:      "servicemeshoperator",
+		Name:      "servicemeshoperator3",
```
The subscription name should probably be a global variable set in one place.

candita commented on test/e2e/gateway_api_test.go:
```diff
@@ -109,54 +108,62 @@ func testGatewayAPIResources(t *testing.T) {
 
 // testGatewayAPIObjects tests that Gateway API objects can be created successfully.
 func testGatewayAPIObjects(t *testing.T) {
-	t.Helper()
```
You could remove the one on L188 too?

Miciah replied:
Yup, `testGatewayAPIResourcesProtection` didn't exist when I made this change.  Fixed in https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5.

candita commented on test/e2e/gateway_api_test.go:
```diff
@@ -265,11 +272,11 @@ func testGatewayAPIResourcesProtection(t *testing.T) {
 			t.Fatalf("failed to find crd %s: %v", crdName, err)
 		}
-		t.Logf("found crd %s at version %s", crdName, crdVersion)
+		t.Logf("Found CRD %s with the following served versions: %s", crdName, strings.Join(crdVersions, ", "))
```
Do you think it might be better to change `assertCRDExists` to make it check for served and storage and ignore the CRD if storage is true?

Also, nit:
```suggestion
		t.Logf("found CRD %s with the following served versions: %s", crdName, strings.Join(crdVersions, ", "))
```

Miciah replied:
Log messages should start with a capital letter by Kubernetes conventions.  Source: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#message-style-guidelines

candita commented on test/e2e/util_gatewayapi_test.go:
```diff
@@ -38,53 +37,27 @@ const (
-// Otherwise returns the CRD version.
-func assertCrdExists(t *testing.T, crdname string) (string, error) {
+// assertCRDExists checks if the CRD of the given name exists and returns an
+// error if not.  If the does CRD exist, this function returns a slice strings
```
```suggestion
// error if not.  If the CRD does exist, this function returns a slice of strings
```

Miciah replied:
Thanks!  Fixed in https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5... and https://github.com/openshift/cluster-ingress-operator/compare/26a3918c7374cc16eac04b87c9e4267591c304a5..f9e552d6aeb31b42bd5791923de4b66f29ec107e (I fixed the "does", then noticed and fixed the "of"..

candita commented on test/e2e/util_gatewayapi_test.go:
```diff
@@ -94,7 +67,7 @@ func assertCrdExists(t *testing.T, crdname string) (string, error) {
 		crdConditions := crd.Status.Conditions
 		for _, version := range crd.Spec.Versions {
 			if version.Served {
```
How about this?  We don't really care about storage versions, right?
```suggestion
			if version.Served  && !version.Storage {
```

Miciah replied:
I'm not sure why we care about the served versions at all in the test output.  However, if it ever became necessary to know, then it seems like it would be more useful to have the full list of served versions.  If anything, it might be useful _also_ to know which one were the stored version.  In any case, I don't want to put in a lot of extra logic in an attempt to anticipate and log exactly the information that could be useful in some hypothetical situation.  (I'm tempted to simplify the logic to simply `t.Logf("%+v", crd)`.)

candita replied:
I don't think we consider either served or stored when we look at CRDs.  Would that pose a problem?

Miciah replied:
I'm not sure what you mean.  Do you mean when the test looks at the CRDs, it doesn't verify that the expected version is served/stored?

candita commented on test/e2e/util_gatewayapi_test.go:
We could make the update corresponding to the TODO in line 546.  Before v1.0.0 the condition didn't exist for Gateway, only GatewayClass.

candita commented on test/e2e/util_gatewayapi_test.go:
```diff
@@ -580,6 +554,7 @@ func assertGatewaySuccessful(t *testing.T, namespace, name string) (*gatewayapiv
 		return false, nil
 	})
 	if err != nil {
+		t.Logf("Last observed gateway:\n%s", util.ToYaml(gw))
 		return nil, fmt.Errorf("gateway %s not %v, last recorded status message: %s", name, gatewayapiv1.GatewayClassConditionStatusAccepted, recordedConditionMsg)
```
As mentioned in https://github.com/openshift/cluster-ingress-operator/pull/1152/files#r1994477437 , we can change `GatewayClassConditionStatusAccepted` to `GatewayConditionStatusAccepted`, to be more correct.

Miciah replied:
Fixed in https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5..

candita commented on test/e2e/util_gatewayapi_test.go:
```diff
@@ -741,121 +698,88 @@ func assertCatalogSource(t *testing.T, namespace, csName string) error {
 	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 30*time.Second, false, func(context context.Context) (bool, error) {
 		if err := kclient.Get(context, nsName, catalogSource); err != nil {
-			t.Logf("failed to get catalogSource %s: %v, retrying...", csName, err)
+			t.Logf("Failed to get CatalogSource %s: %v.  Retrying...", csName, err)
```
Oh, you are using capitalized log messages here?  We don't usually do so.

candita commented on test/e2e/util_gatewayapi_test.go:
```diff
@@ -741,121 +698,88 @@ func assertCatalogSource(t *testing.T, namespace, csName string) error {
-				t.Logf("found ServiceMeshControlPlane %s/%s with ready components: %v", smcp.Namespace, smcp.Name, smcp.Status.Readiness.Components["ready"])
-				return true, nil
-			}
+		if istio.Status.GetCondition(sailv1.IstioConditionReady).Status == metav1.ConditionTrue {
```
Aw, it no longer provides component status?  That was a great troubleshooting tool.

candita commented on test/e2e/util_gatewayapi_test.go:
```diff
@@ -741,121 +698,88 @@ func assertCatalogSource(t *testing.T, namespace, csName string) error {
 				return true, nil
 			}
-			t.Logf("failed to get deleted SMCP %s: %v", nsName.Name, err)
+			t.Logf("Failed to get deleted Istio %s: %v", nsName.Name, err)
```
Is something like this more accurate?
```suggestion
			t.Logf("Deleted Istio %s not yet recreated: %v", nsName.Name, err)
```

Miciah replied:
Maybe something like, "Unexpected error when trying to get Istio after deletion"?

candita replied:
That would work.

Miciah replied:
Done in https://github.com/openshift/cluster-ingress-operator/compare/8055a976678a3377646399d0110f8cbe4d0e9144..25dde982b0f91fe5bed88fc44da904d920f03d2a.

candita commented on test/e2e/util_gatewayapi_test.go:
```diff
@@ -741,121 +698,88 @@ func assertCatalogSource(t *testing.T, namespace, csName string) error {
+		// Compare the UID to determine whether it is the same object
+		// or whether it is a new one with the same name.
+		if newIstio.UID != existingIstio.UID {
+			t.Logf("Istio %s has been recreated (old UID: %v, new UID: %v).", nsName.Name, existingIstio.UID, newIstio.UID)
```
Punctuation?
```suggestion
			t.Logf("Istio %s has been recreated (old UID: %v, new UID: %v)", nsName.Name, existingIstio.UID, newIstio.UID)
```

Miciah replied:
Fixed in https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5.

candita commented on test/e2e/util_gatewayapi_test.go:
You removed `assertDNSRecord`, but I felt it was helpful in waiting for the record to be published.  Is DNS working out ok without it?

Miciah replied:
Ah, I didn't realize that the purpose of `assertDNSRecord` was to introduce a delay for the DNS record to propagate before the test tried to resolve it.  

The logic that looks at the DNSRecord CR seems out of place for this test; DNSRecord is an internal API, not a user-facing API, and certainly not part of the intended workflow.  

https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5 restores `assertDNSRecord`, but we should find a better way to deal with DNS propagation delays (for this test and for others!).  Even with `assertDNSRecord`, I need to add a `time.Sleep` or else I seem to get the cached NXDOMAIN response, and the test times out after the 10-minute polling loop.

candita replied:
It may depend on which build machine this runs. Grant found/fixed an issue in one of the build machines where it was caching errors for a much longer time than needed, and iirc Bear Chan fixed this by changing the `negativeTtl`.  Maybe it was reverted or this is a build machine that still has it.  https://redhat-internal.slack.com/archives/CBN38N3MW/p1725382862844289?thread_ts=1723658374.670269&cid=CBN38N3MW

Miciah commented on pkg/operator/controller/gateway-service-dns/controller_test.go:
```diff
@@ -60,28 +60,23 @@ func Test_Reconcile(t *testing.T) {
-	}
-	exampleGatewayLabel := map[string]string{
-		"istio.io/gateway-name": "example-gateway",
+		"istio.io/gateway-name":    "example-gateway",
```
Yup!  Fixed in https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5.

Miciah commented on pkg/operator/controller/gatewayclass/istio.go:
```diff
@@ -0,0 +1,209 @@
+								Disabled: ptr.To(true),
+							},
+							MetadataExchangeHeaders: &sailv1.ProxyConfigProxyHeadersMetadataExchangeHeaders{
+								Mode: sailv1.ProxyConfigProxyHeadersMetadataExchangeModeInMesh,
```
Yup!  Done in https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5.

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/compare/30b576c4293b2962b137bcb93ff77f278d7bf681..26a3918c7374cc16eac04b87c9e4267591c304a5 includes many updates:

- Updated the list of features for the conformance tests, increased the timeout, and fixed the `sed` command.  
- Added a "gateway labeler" controller to add the `istio.io/rev` label to gateways that specify any gatewayclass with our controller name.  This is necessary for OSSM 3 to reconcile these gateways.  
- Removed the stop-gap hard-coded logic that added the `istio.io/rev` label to the "openshift-ingress" namespace.  
- Fixed a test for the gateway-service-dns controller.  
- Moved some names to `names.go`.  
- Removed the error return value from `desiredIstio` and added more comments.  
- Lot's of tidy-ups in the E2E tests (tweaked log messages and comments, restored `assertDNSRecord`).

Miciah commented:
/test e2e-aws-operator
/test e2e-aws-gatewayapi
/test e2e-aws-operator-techpreview

shaneutt reviewed (commented):
Did a review pass, things are looking good! Most of my comments are prescriptive suggestions or tiny nits (anything I label as `nit:` please always feel free to resolve at your discretion). I do have one open ended one in there which I apologize for, but it might be worth thinking though a bit further.

Miciah commented on hack/gatewayapi-conformance.sh:
```diff
@@ -43,11 +43,11 @@ cd gateway-api
-# because the AWS ELB needs extra ~60s for DNS propagation
-sed -i "s/MaxTimeToConsistency:              30/MaxTimeToConsistency:              90/g" conformance/utils/config/timeout.go
+# Modify the default timeout of "MaxTimeToConsistency" to make tests pass on AWS
+# because the AWS ELB needs an extra ~60s for DNS propagation.
```
Fair enough.  Done in https://github.com/openshift/cluster-ingress-operator/compare/f9e552d6aeb31b42bd5791923de4b66f29ec107e..7a31d446714645793243c51dc58520d05938191b.

shaneutt replied:
:+1: please consider this resolved.

Miciah commented on hack/gatewayapi-conformance.sh:
```diff
@@ -43,11 +43,11 @@ cd gateway-api
-sed -i "s/MaxTimeToConsistency:              30/MaxTimeToConsistency:              90/g" conformance/utils/config/timeout.go
+# Modify the default timeout of "MaxTimeToConsistency" to make tests pass on AWS
+# because the AWS ELB needs an extra ~60s for DNS propagation.
+sed -i -e '/MaxTimeToConsistency:/ s/30/90/' conformance/utils/config/timeout.go
 
-SUPPORTED_FEATURES="Gateway,HTTPRoute,ReferenceGrant,GatewayPort8080,HTTPRouteQueryParamMatching,HTTPRouteMethodMatching,HTTPRouteResponseHeaderModification,HTTPRoutePortRedirect,HTTPRouteSchemeRedirect,HTTPRoutePathRedirect,HTTPRouteHostRewrite,HTTPRoutePathRewrite,HTTPRouteRequestMirror,HTTPRouteRequestMultipleMirrors,HTTPRouteBackendProtocolH2C,HTTPRouteBackendProtocolWebSocket"
+SUPPORTED_FEATURES='Gateway,HTTPRoute,GatewayPort8080,GatewayStaticAddresses,GatewayHTTPListenerIsolation,HTTPRouteBackendRequestHeaderModification,HTTPRouteQueryParamMatching,HTTPRouteMethodMatching,HTTPRouteResponseHeaderModification,HTTPRoutePortRedirect,HTTPRouteSchemeRedirect,HTTPRoutePathRedirect,HTTPRouteHostRewrite,HTTPRoutePathRewrite,HTTPRouteRequestMirror,HTTPRouteRequestMultipleMirror ...
 
 echo "Start Gateway API Conformance Testing"
-go test ./conformance -v -timeout 10m -run TestConformance -args --supported-features=${SUPPORTED_FEATURES}
+go test ./conformance -v -timeout 2h -run TestConformance -args "--supported-features=${SUPPORTED_FEATURES}"
```
I'd add a comment if I knew the reason.  Is it expected that they complete in less than an hour?

shaneutt replied:
When I worked on the conformance tests and reports with Mattia I remember them taking several minutes... could be I'm forgetting something. It seems worth following up on, but doesn't have to hold this up since we have significant time constraints. I recommend creating some sort of follow-up JIRA or issue to look into it later, and then consider this comment resolved.

lihongan replied:
These is blocker PR https://github.com/kubernetes-sigs/gateway-api/pull/3389 for v1.2.1 CRDs conformance test, if that issue (CoreDNS pod crash) get fixed the conformance tests should be completed in 20 minutes.

shaneutt replied:
```suggestion
// This currently takes a long time because of issues with CoreDNS pods crashing during the tests in upstream Gateway API.
// When https://github.com/kubernetes-sigs/gateway-api/pull/3389 gets resolved it should take much less time.
```

lihongan replied:
I ran conformance test manually and found another "timeout" issue, seems the istiod controller was not created until the gatewayclass named `openshift-default` is created explicitly. In 4.18 when creating the gatewayclass `gateway-conformance` the default one will be created automatically, so that is fine. But with this PR, the default one is not created and the gateway stuck in pending state.
One of workaround is changing the gatewayclass name and adding tag `--gateway-class` in go test 
```
oc apply -f -<<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: GatewayClass
metadata:
  name: openshift-default
<...snip...>
go test ./conformance -v -timeout 30m -run TestConformance -args  --gateway-class=openshift-default --supported-features...
``` 
BTW, there are some failed tests in my local run and will investigate later, looks we should increase the timeout to 30m.
```
$ grep FAIL ossm3_test_result.yaml 
--- FAIL: TestConformance (1201.14s)
    --- FAIL: TestConformance/GatewayHTTPListenerIsolation (495.73s)
        --- FAIL: TestConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes (180.00s)
    --- FAIL: TestConformance/GatewayStaticAddresses (2.35s)

```

Miciah commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,203 @@
+package gateway_labeler
+
+import (
+	"context"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	operatorcontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
```
Right, we have discussed import grouping extensively.  At some point we need to decide on a standard and enforce it consistently.  For now, I'm following (and insisting on) a bare minimum of standard libraries first, everything else second (optionally divided into further groups, which we do very inconsistently).

shaneutt replied:
Gotcha, maybe just a follow up JIRA/issue for later for us to take another look at this later. Not important right now. Please consider this comment resolved.

Miciah commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,203 @@
+	gatewayClassToGateways := func(ctx context.Context, o client.Object) []reconcile.Request {
+		requests := []reconcile.Request{}
+
+		// Using an index might be more efficient than listing all
+		// gateways and then filtering.  However, setting up an index is
+		// complicated by the fact that the Gateway API CRDs might not
+		// exist when this controller is initialized.
```
I'm glad I included the comment then!  Yeah, I spent a lot of time writing index code, then seeing it fail to initialize, and eventually I concluded it would be too much effort and too much complexity for a small potential benefit.

shaneutt replied:
Sounds good. Consider this resolved.

Miciah commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,203 @@
+		// exist when this controller is initialized.
+
+		var gateways gatewayapiv1.GatewayList
+		if err := reconciler.cache.List(ctx, &gateways); err != nil {
+			log.Error(err, "Failed to list gateways for gatewayclass", "gatewayclass", o.GetName())
+			return requests
+		}
```
If we're confident that the `List` cannot return an error, then one option would be to panic on error.  (Controller-runtime will recover if the reconciler panics, but not if the mapfunc panics.)  

Alternatively, we could write some complicated logic to make the controller resynch if the mapfunc gets an error.  

As a last resort, the cache [has a configurable sync period](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/cache#Options.SyncPeriod), which defaults to 10 hours.  We could shorten that; a cluster typically shouldn't have _very_ many gateways or gatewayclasses that resynching them more often would be a problem.  I wouldn't want to shorten it _much_ though, and I don't know if eventual consistency after, say, 1 hour is that much better than eventual consistency after 10 hours.  

Personally, I would go with the panic option, or just not worry about it.  It's an extremely unlikely failure mode (as far as we can reason it out), it gets logged, and the workaround is easy: touch the gatewayclass or gateway to trigger a reconcilation.

shaneutt replied:
> Personally, I would go with the panic option, or just not worry about it. It's an extremely unlikely failure mode (as far as we can reason it out), it gets logged, and the workaround is easy: touch the gatewayclass or gateway to trigger a reconcilation.

That sounds fine given our time constraints.

The real fix is resolving https://github.com/kubernetes-sigs/controller-runtime/issues/1996 anyway so that failures just have a re-queue catch in map functions (at least, this is the fix I want :partying_face:).

Miciah commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,203 @@
+
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool { return false },
```
> So _technically_ if a `GatewayClass` gets deleted then there's potentially `Gateway` resources still referring to it which this controller may have labeled. I guess in a perfect world you would want to clean up those labels in case the `Gateway` gets reused (this actually can happen when migrating between implementation, which is something we've prescribed in upstream before).

I'll ask the Service Mesh team what happens if I remove the label once Istio has created the deployment and service for a gateway.  However, I figure that once Istio creates resources for a gateway, you want Istio to keep managing those resources.  You wouldn't want those resources to be orphaned or want them to be deleted without some more explicit action.  Put another way, I want to limit the blast radius of changing an in-use gateway's gatewayclass or the gatewayclass's controller name.  

I alluded to my concern here: https://github.com/openshift/cluster-ingress-operator/blob/5ffbd8b2b5878c2d62d9409c2fb16fcf202cc6cf/pkg/operator/controller/gateway-labeler/controller.go#L114-L120

> In a related, but inverted way: they could also "detach" the `Gateways` by dropping or changing the `parametersRef`. At that point it would seem unreasonable to remove our labels because they've specified we're no longer responsible for the resource. Just one of the cute side effects of using labels in this way, I suppose.

I don't understand that.  Are you saying that the gatewayclass's `parametersRef` could have some implementation-specific configuration that overrides the usual logic that the controller named by the gatewayclass referenced by the gateway owns the gateway?  

> This is an edge-case where the worst effect would be "I migrated my Gateway from one Istio implementation to another, and the new implementation did not pick it up, and I saw a label I didn't put there, and it made me a bit confused". So it would appear to be something we should NOT actually be worrying about right now, but I figured I would mention it while we're here.

I'm thinking it's something to document ("How do I make OpenShift disown my gateway?"), but I dislike that it involves an Istio-ism ("Remove the `istio.io/rev` label!").  

>  I would be a little curious to see what Istio does with orphaned `Gateways` in either of the above mentioned cases though 🤔

+1.

> I think we should add some comments and documentation here that point out for future readers that we _don't_ try to clean up these labels. Other than that, perhaps just some food for thought.

I have the comment I referenced earlier, but I can add another to the gatewayclass's `DeleteFunc`.  Done in https://github.com/openshift/cluster-ingress-operator/compare/f9e552d6aeb31b42bd5791923de4b66f29ec107e..7a31d446714645793243c51dc58520d05938191b.

shaneutt replied:
:+1:, consider this resolved.

Miciah commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,203 @@
+
+			switch {
+			case old.Spec.ControllerName == new.Spec.ControllerName:
+				return false
```
Done in https://github.com/openshift/cluster-ingress-operator/compare/f9e552d6aeb31b42bd5791923de4b66f29ec107e..7a31d446714645793243c51dc58520d05938191b.

Miciah commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,203 @@
+				// If Istio created resources for the gateway,
+				// then we want Istio to continue managing those
+				// resources.  The user can delete the gateway
+				// and related resources if desired.
```
Done in https://github.com/openshift/cluster-ingress-operator/compare/f9e552d6aeb31b42bd5791923de4b66f29ec107e..7a31d446714645793243c51dc58520d05938191b.

Miciah commented on pkg/operator/controller/gatewayclass/controller.go:
```diff
@@ -131,17 +122,19 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
-		r.startSMCPWatch.Do(func() {
-			isOurSMCP := predicate.NewPredicateFuncs(func(o client.Object) bool {
-				return o.GetName() == operatorcontroller.ServiceMeshControlPlaneName(r.config.OperandNamespace).Name
+		// The OSSM operator installs the istios.sailoperator.io CRD.
+		// We must create the watch for this resource only after the
+		// operator is installed.  We use sync.Once here to start the
+		// watch for istios only once.
+		r.startIstioWatch.Do(func() {
```
Are you saying that instead of the following:

- Controller A watches gatewayclasses and reconciles them (if the controller name matches) by installing OSSM, creating an Istio CR, and adding a watch on istios.  

You are proposing the following?

- Controller A watches gatewayclasses and reconciles them by installing OSSM.  
- Controller B watches CRDs and reconciles them by starting controller C when it observes that the Istio CRD exists.  Controller B could also start the gatewayclass, gateway-service-dns, and gateway-labeler controllers when it observed that the Gateway API CRDs were created.
- Controller C watches istios and reconciles them by creating an Istio CR.  

That does make each controller simpler.  It might turn out to be simpler overall.  However, it's a large change, it isn't directly related to the OSSM 3 bump, and so I'd rather leave it for a follow-up.

shaneutt replied:
Yeah, let's leave things alone for now given our time constraints and the fact that this has been working so far. Maybe a follow-up JIRA/issue to take a look at it at some later point would be good. In any case please consider this resolved :+1:

Miciah replied:
I created https://issues.redhat.com/browse/NE-1986 for the follow-up.

Miciah commented on pkg/operator/controller/gatewayclass/istio.go:
```diff
@@ -0,0 +1,210 @@
+	}
+
+	// TODO If we have a current CR with a different owner reference,
+	// should we append the new gatewayclass?
```
> If I'm understanding the question right, this is predicated on the idea that we might find an Istio resource that has an `OwnerReference` on it that we didn't expect.

Right.

> The `Istio` resource here is uniquely named, and in a very specific namespace we control so anything extra that crops up would have to be from:

Istios [are cluster-scoped](https://github.com/istio-ecosystem/sail-operator/blob/1.0.0/api/v1/istio_types.go#L243C32). 

>     1. a cluster admin surgically tampering with resources in the namespace in a way that could only be seen as "extremely deliberate"
> 
>     2. a bug in our own logic
> 
> 
> Right?

Right (except for the namespace).

> If so it would seem reasonable to not append, but if you wanted to know if that condition ever occurs you could error log it.

Yeah.  Note that this is a to-do comment that is carried over from the OSSM 2 ServiceMeshControlPlane logic, but I can go ahead and update it.  Done in https://github.com/openshift/cluster-ingress-operator/pull/1152/commits/4d29932c9f1719e1a92490adfca44a9b609d1379.

shaneutt replied:
:+1: /resolve

Miciah commented on pkg/operator/controller/gatewayclass/istio.go:
```diff
@@ -0,0 +1,210 @@
+					IstioNamespace: ptr.To(controller.DefaultOperandNamespace),
+					// Logging: &sailv1.GlobalLoggingConfig{
+					// 	Level: ptr.To("all:debug"),
+					// },
```
It's kind of vestigial, but it's likely to be useful in the future, so I thought it was worth leaving in.  I can delete it though.  Done in https://github.com/openshift/cluster-ingress-operator/compare/f9e552d6aeb31b42bd5791923de4b66f29ec107e..7a31d446714645793243c51dc58520d05938191b.

shaneutt replied:
That's fine, sg! /resolve

Miciah commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,203 @@
+
+var log = logf.Logger.WithName(controllerName)
+
+// NewUnmanaged creates and returns a controller that adds the "istio.io/rev"
+// label to gateways so that Istio reconciles them.  This is an unmanaged
+// controller, which means that the manager does not start it.
```
Done in https://github.com/openshift/cluster-ingress-operator/compare/f9e552d6aeb31b42bd5791923de4b66f29ec107e..7a31d446714645793243c51dc58520d05938191b.

(I changed "Sail Operator" to "OSSM Operator" per https://github.com/openshift/cluster-ingress-operator/pull/1152#discussion_r1994370215.)

Miciah commented on pkg/operator/controller/gatewayclass/istio.go:
```diff
@@ -0,0 +1,210 @@
+		// haven't really found any explanation for.
+		"PILOT_MULTI_NETWORK_DISCOVER_GATEWAY_API": "false",
+	}
+	istio := sailv1.Istio{
```
Done in https://github.com/openshift/cluster-ingress-operator/compare/f9e552d6aeb31b42bd5791923de4b66f29ec107e..7a31d446714645793243c51dc58520d05938191b.

Miciah commented:
[e2e-aws-operator-techpreview failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1152/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator-techpreview/1900542460317143040) because `TestGatewayAPI/testGatewayAPIObjects` failed:
```text
    util_gatewayapi_test.go:685: GET test-hostname-hkslf.gws.ci-op-hfg9xwxh-9e7c5.origin-ci-int-aws.dev.rhcloud.com failed: status 503, expected 200, retrying...
    gateway_api_test.go:175: error contacting test-hostname-hkslf.gws.ci-op-hfg9xwxh-9e7c5.origin-ci-int-aws.dev.rhcloud.com's endpoint: context deadline exceeded
```
https://github.com/openshift/cluster-ingress-operator/pull/1152/commits/7a31d446714645793243c51dc58520d05938191b adds some logging to help diagnose such failures.

Miciah commented:
/test e2e-aws-operator
/test e2e-aws-gatewayapi
/test e2e-aws-operator-techpreview
/test e2e-aws-gatewayapi-conformance

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/pull/1152/commits/1631ac055c7289c502aa7840716e766468719900 updates `gatewayapi-conformance.sh` to cherry-pick https://github.com/kubernetes-sigs/gateway-api/commit/f64c54a3606c8eee5b4c85b1c5f8f0d3cf3470ca to resolve the issue mentioned in https://github.com/openshift/cluster-ingress-operator/pull/1152#discussion_r1997811757.  

/test e2e-aws-gatewayapi-conformance

Miciah commented:
/test e2e-aws-gatewayapi-conformance

Miciah commented:
Just as a test—

/test unit

Miciah commented:
/test e2e-aws-gatewayapi-conformance

lihongan commented:
/test e2e-aws-operator-techpreview

lihongan commented:
/test e2e-aws-gatewayapi-conformance

lihongan commented:
/test e2e-ibmcloud-operator

just a test

lihongan commented:
> @lihongan: No presubmit jobs available for openshift/cluster-ingress-operator@master

looks the tests with tag `always_run: false` have issues, both `e2e-aws-gatewayapi-conformance` and `e2e-ibmcloud-operator` report the error.

danilo-gemoli commented:
/retest-required

lihongan commented:
/test e2e-aws-gatewayapi-conformance

Miciah commented:
> [1631ac0](https://github.com/openshift/cluster-ingress-operator/commit/1631ac055c7289c502aa7840716e766468719900) updates `gatewayapi-conformance.sh` to cherry-pick [kubernetes-sigs/gateway-api@f64c54a](https://github.com/kubernetes-sigs/gateway-api/commit/f64c54a3606c8eee5b4c85b1c5f8f0d3cf3470ca) to resolve the issue mentioned in [#1152 (comment)](https://github.com/openshift/cluster-ingress-operator/pull/1152#discussion_r1997811757).

https://github.com/openshift/cluster-ingress-operator/compare/1631ac055c7289c502aa7840716e766468719900..c90668cd1743d1d23ef3a20a87c1faf1e733d943 adds a `git fetch` command, which is needed for the cherry-pick command.  

/test e2e-aws-gatewayapi-conformance

Miciah commented:
I'm hoping https://github.com/openshift/cluster-ingress-operator/pull/1152/commits/bb1b3b5e4a3bb60475b121cfec7e0bf74381dc15 will improve the conformance test results.  

/test e2e-aws-gatewayapi-conformance

lihongan commented on hack/gatewayapi-conformance.sh:
```diff
@@ -32,22 +32,30 @@ cd "${CLONE_DIR}"
+sed -i -e '/MaxTimeToConsistency:/ s/30/90/' conformance/utils/config/timeout.go
 
-SUPPORTED_FEATURES="Gateway,HTTPRoute,ReferenceGrant,GatewayPort8080,HTTPRouteQueryParamMatching,HTTPRouteMethodMatching,HTTPRouteResponseHeaderModification,HTTPRoutePortRedirect,HTTPRouteSchemeRedirect,HTTPRoutePathRedirect,HTTPRouteHostRewrite,HTTPRoutePathRewrite,HTTPRouteRequestMirror,HTTPRouteRequestMultipleMirrors,HTTPRouteBackendProtocolH2C,HTTPRouteBackendProtocolWebSocket"
+SUPPORTED_FEATURES='Gateway,HTTPRoute,GatewayPort8080,GatewayStaticAddresses,GatewayHTTPListenerIsolation,HTTPRouteBackendRequestHeaderModification,HTTPRouteQueryParamMatching,HTTPRouteMethodMatching,HTTPRouteResponseHeaderModification,HTTPRoutePortRedirect,HTTPRouteSchemeRedirect,HTTPRoutePathRedirect,HTTPRouteHostRewrite,HTTPRoutePathRewrite,HTTPRouteRequestMirror,HTTPRouteRequestMultipleMirror ...
```
Seems the `HTTPRouteDestinationPortMatching` is missing, and maybe remove `GatewayStaticAddresses` since it is hard to test here.

Miciah replied:
`HTTPRouteDestinationPortMatching` is experimental: https://github.com/kubernetes-sigs/gateway-api/blob/v1.2.1/pkg/features/httproute.go#L100-L104 

I'll remove `GatewayStaticAddresses` for the time being.

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/pull/1152/commits/cae4f651710ca40617baba160183f4da2c77b63a changes the name of the gatewayclass used for conformance tests to be shorter to avoid "must be no more than 63 characters" errors and removes `GatewayStaticAddresses` from `SUPPORTED_FEATURES` as it requires a pre-allocated IP address.

Miciah commented:
/test e2e-aws-gatewayapi-conformance

rhamini3 commented:
pre-merge checks completed on `4.19.0-0.ci.test-2025-03-19-174815-ci-ln-jgcmfkt-latest`

Tested:

1.  dnsRecord, lb-service and istio resource are recreated after being deleted
2. corresponding dnsRecord is updated, added and deleted when updating gateway listeners
3. single httproute connects to 2 gateways and is accessible with both hostnames
4. httproute is orphaned after gateway is deleted and the listener dnsrecord is removed
5. httproute can switch connection to gateways without any errors
6. gateway can handle connections to 2 or more httproutes

candita commented on hack/gatewayapi-conformance.sh:
```diff
@@ -32,22 +34,32 @@ cd "${CLONE_DIR}"
-SUPPORTED_FEATURES="Gateway,HTTPRoute,ReferenceGrant,GatewayPort8080,HTTPRouteQueryParamMatching,HTTPRouteMethodMatching,HTTPRouteResponseHeaderModification,HTTPRoutePortRedirect,HTTPRouteSchemeRedirect,HTTPRoutePathRedirect,HTTPRouteHostRewrite,HTTPRoutePathRewrite,HTTPRouteRequestMirror,HTTPRouteRequestMultipleMirrors,HTTPRouteBackendProtocolH2C,HTTPRouteBackendProtocolWebSocket"
+# Test all standard features except GatewayStaticAddresses, which requires a
+# pre-allocated IP address.
+SUPPORTED_FEATURES='Gateway,HTTPRoute,GatewayPort8080,GatewayHTTPListenerIsolation,HTTPRouteBackendRequestHeaderModification,HTTPRouteQueryParamMatching,HTTPRouteMethodMatching,HTTPRouteResponseHeaderModification,HTTPRoutePortRedirect,HTTPRouteSchemeRedirect,HTTPRoutePathRedirect,HTTPRouteHostRewrite,HTTPRoutePathRewrite,HTTPRouteRequestMirror,HTTPRouteRequestMultipleMirrors,HTTPRouteRequestTimeo ...
```
As far as I can tell, there are still some extended features in this list.  

- GatewayHTTPListenerIsolation (fails)
- HTTPRouteBackendRequestHeaderModification
- HTTPRouteRequestTimeout
- HTTPRouteBackendTimeout
- HTTPRouteParentRefPort

Also, a test called HTTPRouteRedirectPortAndScheme runs (and fails) and it isn't in any of the features list, supported or extended.

I opened https://github.com/kubernetes-sigs/gateway-api/issues/3693 but it's possible I've misunderstood the intent of separating standard and extended features.

Miciah replied:
> As far as I can tell, there are still some extended features in this list.
> 
>     * GatewayHTTPListenerIsolation (fails)
> 
>     * HTTPRouteBackendRequestHeaderModification
> 
>     * HTTPRouteRequestTimeout
> 
>     * HTTPRouteBackendTimeout
> 
>     * HTTPRouteParentRefPort

These all appear to be standard in Gateway API v1.2.1:

https://github.com/kubernetes-sigs/gateway-api/blob/v1.2.1/pkg/features/gateway.go#L77-L79
https://github.com/kubernetes-sigs/gateway-api/blob/v1.2.1/pkg/features/httproute.go#L106-L108
https://github.com/kubernetes-sigs/gateway-api/blob/v1.2.1/pkg/features/httproute.go#L161-L163
https://github.com/kubernetes-sigs/gateway-api/blob/v1.2.1/pkg/features/httproute.go#L166-L168
https://github.com/kubernetes-sigs/gateway-api/blob/v1.2.1/pkg/features/httproute.go#L171-L173

They are all also listed in the Istio 1.24 conformance report as supported features: https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/reports/v1.2.1/istio-istio/experimental-1.24-default-report.yaml

I don't understand why they wouldn't be passing for us.  They aren't experimental, they're standard, and they're supposed to be supported by the version of Istio that we are using.  

That said, I'll remove them from `SUPPORTED_FEATURES` for now in the interest of unblocking this PR.  We can follow up on these conformance test failures separately from this PR, which should remain focused on the OSSM 3 bump itself.  

> Also, a test called HTTPRouteRedirectPortAndScheme runs (and fails) and it isn't in any of the features list, supported or extended.
> 
> I opened [kubernetes-sigs/gateway-api#3693](https://github.com/kubernetes-sigs/gateway-api/issues/3693) but it's possible I've misunderstood the intent of separating standard and extended features.

---

We discussed this offline and decided to revert to the original set of supported features that we used for Gateway API v1.0.0.  This means removing the features you listed, as well as GRPCRoute.  I'll also follow up with the OSSM maintainers in case they have some ideas as to why the GRPCRoute tests aren't passing.

candita replied:
> They are all also listed in the Istio 1.24 conformance report as supported features: https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/reports/v1.2.1/istio-istio/experimental-1.24-default-report.yaml

That report is an _experimental_ report.  Few implementers are running the standard report anymore, but there is https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/reports/v1.2.0/kong-gateway-operator/standard-v1.4.0-expressions-report.yaml and https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/reports/v1.2.0/airlock-microgateway/standard-4.5.0-default-report.yaml

lihongan commented:
/test e2e-aws-gatewayapi-conformance

lihongan commented:
/test e2e-aws-gatewayapi-conformance

different failing tests in each run, looks unstable

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/compare/cae4f651710ca40617baba160183f4da2c77b63a..77cd7e972d4817ba6e8be8a1c2f72f0e7e8647da reverts the conformance tests to the original set that we had before this PR (see https://github.com/openshift/cluster-ingress-operator/pull/1152#discussion_r2005993790).

Miciah commented:
/test e2e-aws-gatewayapi-conformance

candita commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,222 @@
+			new := e.ObjectNew.(*gatewayapiv1.GatewayClass)
+
+			switch {
+			case old.Spec.ControllerName == new.Spec.ControllerName:
```
ControllerName is immutable so we could simplify this.  https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/gatewayclass_types.go#L81-L86

Miciah replied:
That's an interesting point.  I wonder whether there could be any edge cases where controller-runtime could register delete and recreate as an update?  Probably not, and this logic is probably not necessary, but it feels more correct to handle updates just in case something really weird happens.

candita commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,222 @@
+		// exist when this controller is initialized.
+
+		var gateways gatewayapiv1.GatewayList
+		if err := reconciler.cache.List(ctx, &gateways); err != nil {
```
I'm curious about why this uses the general `cache` rather than the `gatewaysCache`?

Miciah replied:
At this point, `reconciler.cache` is the same cache as `gatewaysClass`:

https://github.com/openshift/cluster-ingress-operator/blob/77cd7e972d4817ba6e8be8a1c2f72f0e7e8647da/pkg/operator/controller/gateway-labeler/controller.go#L48-L49

The choice of using `reconciler.cache` instead of `gatewaysCache` to refer to the cache might be related to refactoring during development (at one point, my local changes had `gatewayClassToGateways` defined as a separate function, outside of `NewUnmanaged`, with `reconciler` as a receiver), or it might have just made more sense in my head (after all, it's a logically separate function, and if it were defined outside of `NewUnmanaged`, it couldn't use local variables of `NewUnmanaged`, and it would take `reconciler` as a receiver rather than taking `gatewaysCache` as a receiver or parameter).

candita commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,222 @@
+			Name:      string(gateway.Spec.GatewayClassName),
+		}
+		var gatewayClass gatewayapiv1.GatewayClass
+		if err := reconciler.cache.Get(context.Background(), gatewayClassName, &gatewayClass); err != nil {
```
This also uses `cache` rather than `gatewaysCache`.

Miciah commented:
[e2e-aws-gatewayapi-conformance failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1152/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-gatewayapi-conformance/1902755195314507776):
```text
Gateway API CRD bundle-version: v1.2.1
Create GatewayClass "conformance"
gatewayclass.gateway.networking.k8s.io/conformance created
error: timed out waiting for the condition on gatewayclasses/conformance
```
I'll look more into this later today, but let's see whether it's an ephemeral issue.
/test e2e-aws-gatewayapi-conformance

Miciah commented:
[e2e-aws-operator-techpreview failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1152/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator-techpreview/1902755473363308544) on `TestIngressControllerCustomEndpoints`.  Using search.ci, I found the same failure on another PR, so I filed have [OCPBUGS-53432](https://issues.redhat.com/browse/OCPBUGS-53432) to track the test failures.

/test e2e-aws-operator-techpreview

Miciah commented:
[hypershift-e2e-aks failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1152/pull-ci-openshift-cluster-ingress-operator-master-hypershift-e2e-aks/1902755983432617984) because `TestCreateClusterV2`, `TestCreateCluster`, `TestNodePool/HostedCluster0`, `TestNodePool/HostedCluster2`, and `TestAzureScheduler` failed:
```text
=== CONT  TestCreateClusterV2
    hypershift_framework.go:304: failed to create cluster, tearing down: failed to create infra: failed to assign a1f96423-95ce-4224-ab27-4e3dc72facd4 role to service principal, 1969e247-fe99-492c-8b23-73104c07491f for scope /subscriptions/5f99720c-6823-4792-8a28-69efb0719eea/resourceGroups/create-cluster-v2-fhmpd-vnet-create-cluster-v2-fhmpd : failed to execute az command: exit status 1
--- FAIL: TestCreateClusterV2 (218.18s)
```
```text
=== PAUSE TestCreateCluster
=== CONT  TestCreateCluster
    hypershift_framework.go:304: failed to create cluster, tearing down: failed to create infra: failed to assign 0336e1d3-7a87-462b-b6db-342b63f7802c role to service principal, 6f4e7044-2819-4413-9ab5-cc513cecff3a for scope /subscriptions/5f99720c-6823-4792-8a28-69efb0719eea/resourceGroups/create-cluster-62684-vnet-create-cluster-62684 : failed to execute az command: exit status 1
--- FAIL: TestCreateCluster (228.49s)
```
```text
=== CONT  TestNodePool/HostedCluster0
    hypershift_framework.go:304: failed to create cluster, tearing down: failed to create infra: failed to assign b24988ac-6180-42a0-ab88-20f7382dd24c role to service principal, 33c6fffb-febf-4324-9cef-0f1281f4fff0 for scope /subscriptions/5f99720c-6823-4792-8a28-69efb0719eea/resourceGroups/node-pool-bkfxj-vnet-node-pool-bkfxj : failed to execute az command: exit status 1
    --- FAIL: TestNodePool/HostedCluster0 (227.93s)
```
```text
=== CONT  TestNodePool/HostedCluster2
    hypershift_framework.go:304: failed to create cluster, tearing down: failed to create infra: failed to assign a1f96423-95ce-4224-ab27-4e3dc72facd4 role to service principal, 1969e247-fe99-492c-8b23-73104c07491f for scope /subscriptions/5f99720c-6823-4792-8a28-69efb0719eea/resourceGroups/node-pool-mq8sq-nsg-node-pool-mq8sq : failed to execute az command: exit status 1
    --- FAIL: TestNodePool/HostedCluster2 (232.74s)
```
```text
=== CONT  TestAzureScheduler
    hypershift_framework.go:304: failed to create cluster, tearing down: failed to create infra: failed to assign a1f96423-95ce-4224-ab27-4e3dc72facd4 role to service principal, 1969e247-fe99-492c-8b23-73104c07491f for scope /subscriptions/5f99720c-6823-4792-8a28-69efb0719eea/resourceGroups/azure-scheduler-c6675-vnet-azure-scheduler-c6675 : failed to execute az command: exit status 1
--- FAIL: TestAzureScheduler (244.11s)
```
I didn't find a bug report for this issue, but I found some ongoing work to improve the error message for failing `az` commands: https://github.com/openshift/hypershift/pull/5877 

Meanwhile, let's retry the CI job.  
/test hypershift-e2e-aks

candita commented:
/lgtm
/approve

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/compare/77cd7e972d4817ba6e8be8a1c2f72f0e7e8647da..8055a976678a3377646399d0110f8cbe4d0e9144 rebases on #1200 to resolve a conflict in `gateway_api_test.go`.  

Note that https://github.com/openshift/cluster-ingress-operator/pull/1152/commits/1b248fdba175620d717f4fdc4c7075e499d2ebf8 fell out of the rebase; it is superseded by https://github.com/openshift/cluster-ingress-operator/pull/1200/commits/5c30510e451e30d3ae09040cc6aef1ae19a58c49.

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/compare/8055a976678a3377646399d0110f8cbe4d0e9144..25dde982b0f91fe5bed88fc44da904d920f03d2a addresses a couple of review comments that I missed before:

- Amend the comment for `IngressOperatorOwnedAnnotation` to say that this annotation is currently only intended for newly created subscriptions.  I also split this change out into a separate commit as it isn't strictly related to the OSSM 3 bump.  
- Improve the E2E test log message when deleting an Istio and subsequently trying to get it results in an unexpected error.  

/test e2e-aws-gatewayapi-conformance

shaneutt commented on hack/gatewayapi-conformance.sh:
```diff
@@ -32,22 +34,30 @@ cd "${CLONE_DIR}"
 git clone --branch "${BRANCH}" https://github.com/kubernetes-sigs/gateway-api
 cd gateway-api
+
+if [[ "$BUNDLE_VERSION" = "v1.2.1" ]]; then
```
_technically_ you're going to want to cover anything in `v1.2.x`. Hopefully we don't have any more patches in `v1.2`, but you never know.

Miciah replied:
Ah, I hadn't seen that the proposed backport at https://github.com/kubernetes-sigs/gateway-api/pull/3688 had been closed.  

However, I'm ambivalent about changing this, knowing that there's still a chance that there could be a v1.2.2 with the patch, or a v1.2.2 with some other change that causes the cherry-pick to fail.  

If we bump to v1.2.2+ and find that the cherry-pick is still needed, we can update this test then with the knowledge that it remains applicable and necessary.

shaneutt commented on hack/gatewayapi-conformance.sh:
```diff
@@ -32,22 +34,30 @@ cd "${CLONE_DIR}"
 cd gateway-api
+
+if [[ "$BUNDLE_VERSION" = "v1.2.1" ]]; then
+    echo "Cherry-picking fix for CoreDNS deployment issue in Gateway API v1.2.1"
```
```suggestion
    // This is a temporary workaround until Gateway API v1.3.x.
    // See: https://github.com/kubernetes-sigs/gateway-api/pull/3389
    echo "Cherry-picking fix for CoreDNS deployment issue in Gateway API v1.2.1"
```

Miciah replied:
If I push another change, I'll include this comment.  Otherwise, I'd rather not burn the CI cycles.  It's easy enough to look up the PR from the commit.

shaneutt commented on hack/gatewayapi-conformance.sh:
```diff
@@ -32,22 +34,30 @@ cd "${CLONE_DIR}"
 
-# modify default timeout of "MaxTimeToConsistency" to make tests pass on AWS
-# because the AWS ELB needs extra ~60s for DNS propagation
-sed -i "s/MaxTimeToConsistency:              30/MaxTimeToConsistency:              90/g" conformance/utils/config/timeout.go
+# Modify the default timeout of "MaxTimeToConsistency" to make tests pass on AWS
+# because the AWS ELB needs an extra ~60s for DNS propagation.  See
+# <https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/best-practices-dns.html#DNS_change_propagation>.
+sed -i -e '/MaxTimeToConsistency:/ s/30/90/' conformance/utils/config/timeout.go
 
 SUPPORTED_FEATURES="Gateway,HTTPRoute,ReferenceGrant,GatewayPort8080,HTTPRouteQueryParamMatching,HTTPRouteMethodMatching,HTTPRouteResponseHeaderModification,HTTPRoutePortRedirect,HTTPRouteSchemeRedirect,HTTPRoutePathRedirect,HTTPRouteHostRewrite,HTTPRoutePathRewrite,HTTPRouteRequestMirror,HTTPRouteRequestMultipleMirrors,HTTPRouteBackendProtocolH2C,HTTPRouteBackendProtocolWebSocket"
 
 echo "Start Gateway API Conformance Testing"
-go test ./conformance -v -timeout 10m -run TestConformance -args --supported-features=${SUPPORTED_FEATURES}
+go test ./conformance -v -timeout 10m -run TestConformance -args "--supported-features=${SUPPORTED_FEATURES}" "--gateway-class=${GATEWAYCLASS_NAME}"
```
THERE we go :+1:

shaneutt commented on pkg/operator/controller/gateway-labeler/controller.go:
```diff
@@ -0,0 +1,222 @@
+			return requests
+		}
+
+		for i := range gateways.Items {
+			if gateways.Items[i].Labels[key] == value {
+				continue
+			}
+			if string(gateways.Items[i].Spec.GatewayClassName) != o.GetName() {
+				continue
+			}
+			request := reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: gateways.Items[i].Namespace,
+					Name:      gateways.Items[i].Name,
+				},
+			}
+			requests = append(requests, request)
+		}
```
nit:

```suggestion
		for _, gateway := range gateways.Items {
			if gateway.Labels[key] == value {
				continue
			}
			if string(gateway.Spec.GatewayClassName) != o.GetName() {
				continue
			}
			request := reconcile.Request{
				NamespacedName: types.NamespacedName{
					Namespace: gateway.Namespace,
					Name:      gateway.Name,
				},
			}
			requests = append(requests, request)
		}
```

Miciah replied:
We've been tending towards using the index to avoid copying large values in the case of struct types, especially in hot loops or with large slices.  For example, using the index can yield a significant performance improvement when iterating over thousands of routes.  For the code in question, this might not be a particularly hot loop, and there probably aren't thousands or even hundreds of gateways, but I still prefer not to copy the gateway if we only need a few of its fields.

shaneutt reviewed (approved):
With the follow-ups that were already created, I think things are looking good here.

candita commented:
Hopefully this was a transient error.

>    gateway_api_test.go:147: failed to find expected Istio operator: error finding deployment openshift-operators/servicemesh-operator3: context deadline exceeded

/test e2e-aws-gatewayapi

candita commented:
/retest

candita commented:
/retest-required

melvinjoseph86 commented:
/retest

Miciah commented:
/retest-required

Miciah commented:
/retest-required

Miciah commented:
An infrastructure issue (error creating role assignments because we hit a limit) is causing at least the hypershift-e2e-aks failures.  I'm not sure about e2e-hypershift.  The issue is being mitigated (by deleting some old role assignments).  

/retest-required

Miciah commented:
I believe the e2e-hypershift test failures were caused by [OCPBUGS-53462](https://issues.redhat.com/browse/OCPBUGS-53462).  

The hypershift-e2e-aks job failed again.  This time I don't see errors about role assignments, but `TestNodePool/HostedCluster0/Main/TestNTOMachineConfigGetsRolledOut` failed.  The hypershift-e2e-aks job passed on #1204, which includes all the commits form this PR, and I found [OCPBUGS-43824](https://issues.redhat.com/browse/OCPBUGS-43824), so it seems that the test is flaky.  

/test hypershift-e2e-aks

candita commented:
/approve

candita commented:
Sorry, I didn't realize you added another commit.
/unapprove

candita commented:
/cancel approve

candita commented:
/hold

candita commented:
/lgtm

lihongan commented:
/label qe-approved

Miciah commented:
/label acknowledge-critical-fixes-only

Per TRT, "Feature gated code is free to use this label."
